shell/platform/linux/fl_text_input_plugin.cc

```diff
@@ -0,0 +1,289 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/shell/platform/linux/fl_text_input_plugin.h"
+
+#include "flutter/shell/platform/common/cpp/text_input_model.h"
+#include "flutter/shell/platform/linux/public/flutter_linux/fl_json_method_codec.h"
+#include "flutter/shell/platform/linux/public/flutter_linux/fl_method_channel.h"
+
+#include <gtk/gtk.h>
+
+static constexpr char kChannelName[] = "flutter/textinput";
+
+static constexpr char kSetClientMethod[] = "TextInput.setClient";
+static constexpr char kShowMethod[] = "TextInput.show";
+static constexpr char kSetEditingStateMethod[] = "TextInput.setEditingState";
+static constexpr char kClearClientMethod[] = "TextInput.clearClient";
+static constexpr char kHideMethod[] = "TextInput.hide";
+static constexpr char kUpdateEditingStateMethod[] =
+    "TextInputClient.updateEditingState";
+static constexpr char kPerformActionMethod[] = "TextInputClient.performAction";
+
+static constexpr char kInputActionKey[] = "inputAction";
+static constexpr char kTextKey[] = "text";
+static constexpr char kSelectionBaseKey[] = "selectionBase";
+static constexpr char kSelectionExtentKey[] = "selectionExtent";
+static constexpr char kSelectionAffinityKey[] = "selectionAffinity";
+static constexpr char kSelectionIsDirectionalKey[] = "selectionIsDirectional";
+static constexpr char kComposingBaseKey[] = "composingBase";
+static constexpr char kComposingExtentKey[] = "composingExtent";
+
+static constexpr char kTextAffinityDownstream[] = "TextAffinity.downstream";
+
+static constexpr int64_t kClientIdUnset = -1;
+
+struct _FlTextInputPlugin {
+  GObject parent_instance;
+
+  FlMethodChannel* channel;
+
+  // Client ID provided by Flutter to report events with
+  int64_t client_id;
+
+  // Input action to perform when enter pressed
+  gchar* input_action;
+
+  // Input method
+  GtkIMContext* im_context;
+
+  flutter::TextInputModel* text_model;
+};
+
+G_DEFINE_TYPE(FlTextInputPlugin, fl_text_input_plugin, G_TYPE_OBJECT)
+
+// Helper function to just check if a method returned an error
+static gboolean finish_method(GObject* object,
+                              GAsyncResult* result,
+                              GError** error) {
+  g_autoptr(FlMethodResponse) response = fl_method_channel_invoke_method_finish(
+      FL_METHOD_CHANNEL(object), result, error);
+  if (response == nullptr)
+    return FALSE;
+  return fl_method_response_get_result(response, error) != nullptr;
+}
+
+// Called when a response is received from TextInputClient.updateEditingState()
+static void update_editing_state_response_cb(GObject* object,
+                                             GAsyncResult* result,
+                                             gpointer user_data) {
+  g_autoptr(GError) error = nullptr;
+  if (!finish_method(object, result, &error))
```

Comment:
Multi-line conditional bodies require braces by style.

Reply:
Is it considered multi-line because the formatter wrapped it?

Reply:
The rule is based on actual text lines, not statements, so yes.